### PR TITLE
Rework the schema for reusing environments

### PR DESCRIPTION
**Pull request by alalazo**
Refers to https://github.com/spack/spack/pull/45139#issuecomment-2450843931

Currently, the schema reads:
```yaml
  from:
    - type: 
        environment: path_or_name
```
but this can't be extended easily to other types, e.g. to buildcaches, without duplicating the extension keys. Use instead:
```yaml
  from:
    - type: environment 
      path: path_or_name
```

### CODE REVIEW

**Comment by kwryankrattiger** [comment]
My thought on this was you have the "environment" type and then you have the "specific environment" type. The goal with the schema was to make it clear that the two types of reuse using the environment label did two somewhat different things.

I also didn't want to confuse the use of `name` for the other types of reuse, where that field would be allowed but unused.

**Comment by kwryankrattiger** [comment]
If we go this way, I think `ref` should be something else. My first thought is a git ref here, and considering the other fields at this level are related to specs, which could specify a git ref, it is a bit confusing.

Some possible alternatives to make it a bit more clear:
* `name_or_path: <name_or_path>`
* `environment_name: <name_or_path>`
* `env: <name_or_path>`

Or some that are similar to the `include` keyword used for specs. (no exclude...I think that would be annoying)
* `included_environments: [...]` and `external_environmetns: [...]`
* `environments: [...]`

**Comment by alalazo** [comment]
> The goal with the schema was to make it clear that the two types of reuse using the environment label did two somewhat different things.

We can do that without encoding some information in the key of a nested object, because at that point:
1. It's not clear what should be added to the root object or to the level below
2. To have a similar semantic on another source of reuse we have to redefine a new object to the schema

For the first point consider this:
```yaml
from:
- type:
    environment: path_or_name
  include: [ ... ]
```
Why should we require users to write following a `>` pattern? :slightly_smiling_face: On the other hand, it doesn't make much sense to duplicate everything and allow also `include` in the nested dict, and have just `type:` in the upper level.

For the second point consider `buildcaches`:
```yaml
from:
  - type: 
      buildcache: path_or_name
```
or having scope selection for externals:
```yaml
from:
  - type: 
      external: path_or_scope
```
In both cases we have at least to cut and paste definitions in the schema because the "key" name changes.

**TL;DR** Imo the `type` attribute should stay an `enum`, and we shouldn't encode "value" semantic in a dictionary "key".   If we want more checks on each `from:` item we can encode them in a validation function, or in an audit.

**Comment by kwryankrattiger** [comment]
>  If we want more checks on each from: item we can encode them in a validation function, or in an audit.

I think that is fine. My only other gripe on this is the key name `ref` for the name of the environment.

As you pointed out, there are other cases where we will eventually probably want to have a value denoting a specific `buildcache` or `external` path/name. Unifying these is probably a good idea in some sense. But also, the environment includes can have additional caveats to them, so it may be beneficial to have more specific denotations in the schema for *where* an environment reuse is coming from.

**Comment by alalazo** [comment]
> My only other gripe on this is the key name ref for the name of the environment.

Yeah, I'm not tied to the name. But I think long names don't belong to configuration, so I would exclude `name_or_path` and similar. I'd also exclude, if possible, anything that can be used just by environment.

What about using `path` and document that it accepts also a name, or viceversa? More generic names I can think of are not really fully satisfying (e.g. `handle`, or spelling `reference` out).

**Comment by kwryankrattiger** [comment]
I don't hate `path`. And validating that in the config audit is probably fine, I didn't think about that option.

**Comment by alalazo** [comment]
Ok, let me change `ref` to `path` then.

**Review by kwryankrattiger — APPROVED** [review]
LGTM, thanks!

**Review by becker33 — APPROVED** [review]
